Sources/BSON/Document.swift

```diff
@@ -0,0 +1,153 @@
+import Foundation
+import NIO
+
+/// A struct representing the BSON document type.
+@dynamicMemberLookup
+public struct Document {
+    /// The element type of a document: a tuple containing an individual key-value pair.
+    public typealias KeyValuePair = (key: String, value: BSON)
+
+    private var _buffer: ByteBuffer
+
+    /// An unordered set containing the keys in this document.
+    private var keySet: Set<String>
+
+    internal init(_ elements: [BSON]) { fatalError("Unimplemented") }
+
+    internal init(keyValuePairs: [(String, BSON)]) { fatalError("Unimplemented") }
+
+    /// Initializes a new, empty `Document`.
+    public init() { fatalError("Unimplemented") }
+
+    /**
+     * Initializes a new `Document` from the provided BSON data. If validate is `true` (the default), validates that
+     * the data is specification-compliant BSON.
+     *
+     * - Throws:
+     *   - `InvalidArgumentError` if the data passed is invalid BSON
+     *
+     * - SeeAlso: http://bsonspec.org/
+     */
+    public init(fromBSON bson: Data, validate: Bool = true) throws { fatalError("Unimplemented") }
```

Comment:
Now that I'm looking at this signature, I wonder if it would be better to have the validation separated into two separate initializers. In order for someone to call this in a non-`throws` context, they'd need to invoke it like:
```swift
try! Document(fromBSON: data, validate: false) // never fails, but requires try still
```
which is a little verbose.

Reply:
that's a good point. we can't have two identical signatures where only one throws, so I guess we could disambiguate them by using different argument labels. 
I feel like the "default" one should be validating and then we have some explicit label indicating we are not validating in the other case. something like
```swift
init(fromBSON bson: Data) throws

// only use this if you know your BSON data is already valid
init(fromValidBSON bson: Data)
```

don't love the naming here though, any ideas?

Reply:
another possible option is `withoutValidating:`. We could also use a static method and do `Document.fromBSONWithoutValidation(_ bson: Data)`. That said, I don't necessarily like either of these more than `fromValidBSON`.

Another question I have is whether we expect users to make use of this, or whether it's mainly for us in the driver. If it's the latter, we could internalize it and name it whatever we want.

Reply:
I'd be fine with keeping it internal and making it public later if someone asks for it.
we only have a validating version of the init right now after all, and no one has complained. we just don't do as much validation in it as we could do

Reply:
sgtm

Reply:
@mbroadst @nbbeeken does that sound reasonable to you?

Reply:
I think the validating form only as the public one sounds good. We still have to disambiguate though right? We can push the less aesthetically pleasant naming to the internal one

Reply:
ping @mbroadst, would be good to get this merged in the next couple days so we can make progress with Neal's PRs

Reply:
SGTM

Reply:
cool I updated the initializers, I think @nbbeeken can just add the internal versions when he needs them

Reply:
sgtm 🚀